src/backend/distributed/planner/multi_explain.c

```diff
@@ -639,6 +682,192 @@ BuildRemoteExplainQuery(char *queryString, ExplainState *es)
 }
 
 
+bool
+ShouldSaveQueryExplain(QueryDesc *queryDesc, int eflags)
+{
+	PlannedStmt *plannedStmt = queryDesc->plannedstmt;
+
+	return SaveTaskExplainPlans &&
+		   ExecutorLevel == 0 &&
+		   !IsParallelWorker() &&
+		   (eflags & EXEC_FLAG_EXPLAIN_ONLY) == 0 &&
+		   !IsCitusPlan(plannedStmt->planTree);
+}
+
+
+void
+SaveQueryExplain(QueryDesc *queryDesc)
+{
+	/*
+	 * Make sure stats accumulation is done.  (Note: it's okay if several
+	 * levels of hook all do this.)
+	 */
+	InstrEndLoop(queryDesc->totaltime);
+
+	ExplainState *es = NewExplainState();
+
+	es->analyze = true;
+	es->verbose = TaskExplainOptions.verbose;
+	es->buffers = TaskExplainOptions.buffers;
+	es->timing = TaskExplainOptions.timing;
+	es->summary = TaskExplainOptions.summary;
+	es->format = TaskExplainOptions.format;
+	es->costs = TaskExplainOptions.costs;
+
+	ExplainBeginOutput(es);
+
+	if (TaskExplainOptions.query)
+	{
+		ExplainQueryText(es, queryDesc);
+	}
+
+	ExplainPrintPlan(es, queryDesc);
+
+	if (es->costs)
+	{
+		ExplainPrintJITSummary(es, queryDesc);
+	}
+
+	ExplainEndOutput(es);
+
+	MemoryContext oldContext = MemoryContextSwitchTo(TopMemoryContext);
+
+	SavedExplainPlan = makeStringInfo();
+	appendStringInfoString(SavedExplainPlan, es->str->data);
+
+	MemoryContextSwitchTo(oldContext);
+
+	pfree(es->str->data);
```

Comment:
It seems like an oversight that postgres skips `pfree(es)` in `backend/commands/explain.c` if they're going to go as far as `pfree(es->str->data)`

Reply:
Agreed. It might be that the assumption is `es->str->data` might be huge and we don't want to keep it until transaction end but, rest of members in `es` are relatively small and properly freeing them wasn't worth the code complexity. I am keeping it similar to auto_explain's `explain_ExecutorEnd()`.